File: src/rag/LocalDocumentStore.ts
/**
 * LocalDocumentStore - Shared utility for loading and searching local
 * text documents. Used internally by the concrete RAG module implementations.
 *
 * This implements a keyword-based relevance search over local .txt files.
 * In future phases this can be replaced with vector embeddings and a
 * vector database without changing the RAG module interface.
 */

import * as fs from "fs";
import * as path from "path";
import { RetrievalResult } from "./BaseRAGModule";

interface Document {
  filename: string;
  content: string;
  /** Lowercased, whitespace-normalized content for search */
  normalizedContent: string;
  /** Individual sections split by double newlines */
  sections: string[];
}

export class LocalDocumentStore {
  private documents: Document[] = [];
  private readonly dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
    this.loadDocuments();
  }

  /**
   * Load all .txt files from the data directory into memory.
   */
  private loadDocuments(): void {
    if (!fs.existsSync(this.dataDir)) {
      console.warn(`[LocalDocumentStore] Directory not found: ${this.dataDir}`);
      return;
    }

    const files = fs.readdirSync(this.dataDir).filter((f) => f.endsWith(".txt"));

    for (const file of files) {
      const filePath = path.join(this.dataDir, file);
      const content = fs.readFileSync(filePath, "utf-8");
      const normalizedContent = content.toLowerCase().replace(/\s+/g, " ");

      // Split document into sections by double newlines for granular retrieval
      const sections = content
        .split(/\n\n+/)
        .map((s) => s.trim())
        .filter((s) => s.length > 0);

      this.documents.push({
        filename: file,
        content,
        normalizedContent,
        sections,
      });
    }

    console.log(
      `[LocalDocumentStore] Loaded ${this.documents.length} documents from ${this.dataDir}`
    );
  }

  /**
   * Search for relevant sections across all documents using keyword
   * matching with TF-based scoring.
   *
   * @param query - User query to search for
   * @param topK - Number of top results to return
   * @returns Ranked retrieval results
   */
  search(query: string, topK: number = 3): RetrievalResult[] {
    const queryTerms = query
      .toLowerCase()
      .replace(/[^a-z0-9\s]/g, "")
      .split(/\s+/)
      .filter((t) => t.length > 2); // Skip very short tokens

    if (queryTerms.length === 0) {
      return [];
    }

    const scoredResults: RetrievalResult[] = [];

    for (const doc of this.documents) {
      for (const section of doc.sections) {
        const normalizedSection = section.toLowerCase().replace(/\s+/g, " ");
        let score = 0;
        let matchedTerms = 0;

        for (const term of queryTerms) {
          // Count occurrences of each query term in the section
          const regex = new RegExp(term, "gi");
          const matches = normalizedSection.match(regex);
          if (matches) {
            matchedTerms++;
            // Score based on frequency, capped to avoid single-term dominance
            score += Math.min(matches.length, 3);
          }
        }

        // Bonus for matching multiple distinct query terms
        if (matchedTerms > 1) {
          score *= 1 + matchedTerms * 0.3;
        }

        // Normalize score to 0-1 range (approximate)
        const normalizedScore = Math.min(score / (queryTerms.length * 3), 1.0);

        if (normalizedScore > 0.05) {
          scoredResults.push({
            content: section,
            source: doc.filename,
            score: Math.round(normalizedScore * 1000) / 1000,
          });
        }
      }
    }

    // Sort by score descending and return top K
    return scoredResults
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);
  }

  /**
   * Returns the count of loaded documents.
   */
  getDocumentCount(): number {
    return this.documents.length;
  }
}

File: src_python/rag/local_document_store.py
"""
LocalDocumentStore - Shared utility for loading and searching local
text documents. Used internally by the concrete RAG module implementations.

This implements a keyword-based relevance search over local .txt files.
In future phases this can be replaced with vector embeddings and a
vector database without changing the RAG module interface.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass, field

from .base_rag_module import RetrievalResult


@dataclass
class _Document:
    filename: str
    content: str
    # Lowercased, whitespace-normalized content for search
    normalized_content: str
    # Individual sections split by double newlines
    sections: list[str]


class LocalDocumentStore:
    def __init__(self, data_dir: str):
        self._data_dir = data_dir
        self._documents: list[_Document] = []
        self._load_documents()

    def _load_documents(self) -> None:
        """Load all .txt files from the data directory into memory."""
        if not os.path.isdir(self._data_dir):
            print(f"[LocalDocumentStore] Directory not found: {self._data_dir}")
            return

        files = sorted(
            f for f in os.listdir(self._data_dir) if f.endswith(".txt")
        )

        for filename in files:
            file_path = os.path.join(self._data_dir, filename)
            with open(file_path, "r", encoding="utf-8") as fh:
                content = fh.read()

            normalized_content = re.sub(r"\s+", " ", content.lower())

            # Split document into sections by double newlines for granular retrieval
            sections = [
                s.strip()
                for s in re.split(r"\n\n+", content)
                if s.strip()
            ]

            self._documents.append(
                _Document(
                    filename=filename,
                    content=content,
                    normalized_content=normalized_content,
                    sections=sections,
                )
            )

        print(
            f"[LocalDocumentStore] Loaded {len(self._documents)} documents "
            f"from {self._data_dir}"
        )

    def search(self, query: str, top_k: int = 3) -> list[RetrievalResult]:
        """
        Search for relevant sections across all documents using keyword
        matching with TF-based scoring.

        Args:
            query: User query to search for
            top_k: Number of top results to return

        Returns:
            Ranked retrieval results
        """
        # Tokenize query: lowercase, remove non-alphanumeric, skip short tokens
        query_terms = [
            t
            for t in re.sub(r"[^a-z0-9\s]", "", query.lower()).split()
            if len(t) > 2
        ]

        if not query_terms:
            return []

        scored_results: list[RetrievalResult] = []

        for doc in self._documents:
            for section in doc.sections:
                normalized_section = re.sub(r"\s+", " ", section.lower())
                score = 0
                matched_terms = 0

                for term in query_terms:
                    # Count occurrences of each query term in the section
                    matches = re.findall(re.escape(term), normalized_section, re.IGNORECASE)
                    if matches:
                        matched_terms += 1
                        # Score based on frequency, capped to avoid single-term dominance
                        score += min(len(matches), 3)

                # Bonus for matching multiple distinct query terms
                if matched_terms > 1:
                    score *= 1 + matched_terms * 0.3

                # Normalize score to 0-1 range (approximate)
                normalized_score = min(score / (len(query_terms) * 3), 1.0)

                if normalized_score > 0.05:
                    scored_results.append(
                        RetrievalResult(
                            content=section,
                            source=doc.filename,
                            score=round(normalized_score * 1000) / 1000,
                        )
                    )

        # Sort by score descending and return top K
        scored_results.sort(key=lambda r: r.score, reverse=True)
        return scored_results[:top_k]

    def get_document_count(self) -> int:
        """Returns the count of loaded documents."""
        return len(self._documents)
